django/core/management/commands/optimizemigration.py

```diff
@@ -0,0 +1,111 @@
+            )
+
+        # Find migration
+        try:
+            migration = loader.get_migration_by_prefix(app_label, migration_name)
+        except AmbiguityError:
+            raise CommandError(
+                "More than one migration matches '%s' in app '%s'. Please be "
+                "more specific." % (migration_name, app_label)
+            )
+        except KeyError:
+            raise CommandError(
+                "Cannot find a migration matching '%s' from app '%s'." %
+                (migration_name, app_label)
+            )
+
+        # Optimize migration
+        optimizer = MigrationOptimizer()
+        new_operations = optimizer.optimize(migration.operations, migration.app_label)
+
+        if len(migration.operations) == len(new_operations):
+            if verbosity > 0:
+                self.stdout.write("  No optimizations possible.")
+            return
+        else:
+            if verbosity > 0:
+                self.stdout.write(
+                    "  Optimizing from %d operations to %d operations." %
+                    (len(migration.operations), len(new_operations))
+                )
+            if check:
+                sys.exit(1)
+
+        # Set the new migration optimizations
+        migration.operations = new_operations
+
+        # Write out the optimized migration file
+        writer = MigrationWriter(migration)
+        migration_file_string = writer.as_string()
+        if writer.needs_manual_porting:
+            if migration.replaces:
+                raise CommandError(
+                    'Migration will require manual porting but is already a squashed migration.\n'
+                    'Transition to a normal migration first: '
+                    'https://docs.djangoproject.com/en/%s/topics/migrations/#squashing-migrations' % get_docs_version()
+                )
+            # Make a new migration with those operations
+            subclass = type("Migration", (migrations.Migration,), {
+                "dependencies": migration.dependencies,
+                "operations": new_operations,
+                "replaces": [(migration.app_label, migration.name)],
+            })
+            optimized_migration_name = '%s_optimized' % migration.name
+            optimized_migration = subclass(optimized_migration_name, app_label)
+            writer = MigrationWriter(optimized_migration)
+            migration_file_string = writer.as_string()
+            if verbosity > 0:
+                self.stdout.write(
+                    self.style.MIGRATE_HEADING('Manual porting required') + '\n'
+                    '  Your migrations contained functions that must be manually copied over,\n'
+                    '  as we could not safely copy their implementation.\n'
+                    '  See the comment at the top of the optimized migration for details.'
+                )
+
+        with open(writer.path, "w", encoding="utf-8") as fh:
+            fh.write(migration_file_string)
+
```

Comment:
We need to `run_formatters()` on new files, see d113b5a837f726d1c638d76c4e88445e6cd59fd5.